### PR TITLE
All routes should be specified as 16-bit routes.

### PR DESCRIPTION
**Pull request by jenshz**
While trying out CSP 2.0, I noticed that one of my nodes that had a routing table didn't work anymore.
We would like to make the upgrade to CSP2.0 as painless as possible, however there is the challenge of having a routing table already defined.

With this change,

`set csp_rtable '0/0 CAN0, 24/3 KISS, 29/5 TXS'`

becomes

`set csp_rtable '0/0 CAN0, 24/14 KISS, 29/16 TXS'`

This route will then both work for CSP 2.0 and CSP 1.0. Otherwise the route would have to be specified in a different way for the two versions (in addition to some more code changes, if CSP1.0 compatibility with the old route definition was to be preserved).

### CODE REVIEW

File: include/csp/arch/csp_thread.h [hunk 1]
```diff
@@ -90,6 +90,9 @@ typedef csp_thread_return_t (* csp_thread_func_t)(void *) __attribute__((stdcall
 */
 #if (CSP_FREERTOS)
 
+#include <FreeRTOS.h>
+#include <task.h>
```
Comment (johandc) [inline include/csp/arch/csp_thread.h:94]: This should not be needed. The idea of /arch is specifically not to include platform specific headers

File: meson.build [hunk 2]
```diff
@@ -63,6 +63,8 @@ csp_inc = include_directories(['.', 'include'])
 install_subdir('include', install_dir : '.')
 install_headers(csp_config_h, install_dir : 'include/csp')
 
+csp_deps += dependency('freertos', fallback: ['freertos', 'freertos_dep'], required: false)
```
Comment (johandc) [inline meson.build]: Why add this here, the FreeRTOS dep is included by the meson build file in sec/arch/FreeRTOS/meson build

File: src/arch/freertos/csp_semaphore.c [hunk 3]
```diff
@@ -20,6 +20,8 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 
 #include <csp/arch/csp_semaphore.h>
 #include <csp/csp_debug.h>
+#include <FreeRTOS.h>
```
Comment (johandc) [inline src/arch/freertos/csp_semaphore.c:23]: Yes, I agree this should be added

File: meson.build [hunk 4]
```diff
@@ -63,7 +63,7 @@ csp_inc = include_directories(['.', 'include'])
 install_subdir('include', install_dir : '.')
 install_headers(csp_config_h, install_dir : 'include/csp')
 
-csp_deps += dependency('freertos', fallback: ['freertos', 'freertos_dep'], required: false)
+csp_deps += dependency('freertos', required: false)
```
Comment (johandc) [inline meson.build:66]: Why would you remove the fallback? That requires FreeRTOS to be installed in the system path, and be found by the libtool, which I think is quite rare. The fallback enabled the possibility to have FreeRTOS as another meson subproject.

File: meson.build (hunk 2 above)
Comment (jenshz) [reply meson.build]: I couldn't get it to compile otherwise. Try compiling on C21.

**Comment by jenshz** [comment]
> Why is this better? It does not reuse the incoming packet for the response. And thus takes up two buffers instead of one.

csp_sendto_reply will modify the packet, thus the second transmission will be sent to ourselves (a reply to a reply).
This at least changes that so we don't do something that's completely wrong. Wasting one buffer allocation in the call, for something that's clearly debugging oriented and only meant to be run sparingly, seemed worth the extra cost of making it more "correct".

**Comment by jenshz** [comment]
> I don't understand the rationale for this. The idea of having backwards compatibility with CSP 1 is to not require an update of the node numbers and routing tables. This breaks that and is somewhat in between CSP 1 and 2.

Please try the CSP 1 compatibility mode with the masks and you'll see why this change matters.
I agree that it's not completely ideal, but otherwise we need to have dynamic calls to get the current netmask length everywhere.

File: meson.build (hunk 4 above)
Comment (jenshz) [reply meson.build:66]: It compiles just fine with my version of meson (0.54.2) with the subprojects using this syntax.
It didn't work if you don't have an explicit FreeRTOS dependency however, breaking compatibility with our Zynq project (that has FreeRTOS hidden under another dependency and not explicitly providing a FreeRTOS dependency).

File: include/csp/arch/csp_thread.h (hunk 1 above)
Comment (jenshz) [reply include/csp/arch/csp_thread.h:94]: There's a `#include <Windows.h>` a couple of lines further up, so that's a pretty moot point?
Please add a comment explaining this for future developers then :)

**Comment by jenshz** [comment]
I think your comments are valid - you can pick and choose the bug-fixes you need for upstream if needed.

File: include/csp/arch/csp_thread.h (hunk 1 above)
Comment (johandc) [reply include/csp/arch/csp_thread.h:94]: Yeah, I don't have anything to do with the windows port. So I also don't see why that was needed there.

File: meson.build (hunk 2 above)
Comment (johandc) [reply meson.build]: We do compile for c21 all the time. Did you remember to set the OS to FreeRTOS? So the right meson.buils include file is included?

File: meson.build (hunk 4 above)
Comment (johandc) [reply meson.build:71]: 1. According to the meson docs, It shouldn't be working without the fallback. The reason why your build does work, must be because FreeRTOS is found somewhere in the search path.

2. It is against the meson philosophy to hide libraries in other libraries. But yes, that is how it is done within the zynq BSP. I think we somewhere have a dependency on either freertos or bsp in order to include the headers from the BSP instead, as a workaround. 

But anyway, I cannot see why removing the fallback should fix any of that.?

**Comment by johandc** [comment]
> > Why is this better? It does not reuse the incoming packet for the response. And thus takes up two buffers instead of one.
> 
> csp_sendto_reply will modify the packet, thus the second transmission will be sent to ourselves (a reply to a reply).
> This at least changes that so we don't do something that's completely wrong. Wasting one buffer allocation in the call, for something that's clearly debugging oriented and only meant to be run sparingly, seemed worth the extra cost of making it more "correct".

Okay. I see now. I just haven't come across any cases where sendto_reply is used more than once. Because multiple packets are typically not used as connectionless.

**Comment by johandc** [comment]
> > I don't understand the rationale for this. The idea of having backwards compatibility with CSP 1 is to not require an update of the node numbers and routing tables. This breaks that and is somewhat in between CSP 1 and 2.
> 
> Please try the CSP 1 compatibility mode with the masks and you'll see why this change matters.

I'm not sure I understand what you mean by "with the masks"? When CSP 1 is used the routing table must be given as it has always been given. With /5 signifying a single node.
With CSP 2 it must use extended mask, so now /14 signifies the single node.

We don't expect people will want to move dynamically between CSP versions, hence why compatibility with the old routing table format is preferred over, the flexibility of swapping protocol without routing table.

> I agree that it's not completely ideal, but otherwise we need to have dynamic calls to get the current netmask length everywhere.

Yes. Dynamic call to get mask length. That's how I implemented it. Just a single function call. And it is only really used in the routing table parser. So nothing too harmful. So again, I believe we prefer to keep full backwards compatibility. I.e. if you upgrade to CSP 2.0 master, it just works as a CSP 1.6. untill you set the config.version = 2. (And provide a new routing table)

File: meson.build [hunk 5]
```diff
@@ -19,6 +19,11 @@ conf.set('CSP_USE_DEDUP', get_option('use_dedup') == true ? 1 : 0)
 
 conf.set('CSP_HAVE_STDBOOL_H', cc.has_header('stdbool.h') == true ? 1 : 0)
 
+zmq_dep = dependency('libzmq', version: '>4.0', required: false)
+if zmq_dep.found()
```
Comment (johandc) [inline meson.build:23]: Again, why is this needed here?

We have this

zmq_dep = dependency('libzmq', version: '>4.0', required: false)
if zmq_dep.found()
	csp_sources += files('csp_if_zmqhub.c')
	csp_deps += zmq_dep
endif

Already in sec/interfaces/meson build
